Safer modify new coins

morcos opened the pull request:
This makes ModifyNewCoins more conservative in when it marks newly created outputs as FRESH.  The usage of ModifyNewCoins in the code base doesn't actually require this because its only ever used on a new cache, however I think its dangerous code that is just asking for a consensus mistake to be used in the future.  The comments in the code hopefully clarify a bit better the proper assumptions that need to be made going forward.


EDIT: The changes to the unit test are a bit complicated, but they demonstrate the problem.  They fail without the prior commit.

morcos committed:
Fix dangerous condition in ModifyNewCoins.

We were marking coins FRESH before being sure they were not overwriting dirty undo data. This condition was never reached in existing code because undo data was always flushed before UpdateCoins was called with new transactions, but could have been exposed in an otherwise safe refactor.
Clarify in the comments the assumptions made in ModifyNewCoins.
Add ability to undo transactions to UpdateCoins unit test.
Thanks to Russ Yanofsky for suggestion on how to make logic clearer and fixing up the ccoins_modify_new test cases.

laanwj commented:
Concept ACK, thanks for improving documentation around this code.

ryanofsky commented on src/test/coins_tests.cpp:
```diff
@@ -227,77 +247,139 @@ BOOST_AUTO_TEST_CASE(updatecoins_simulation_test)
+            // Remove new outputs
+            result[undohash].Clear();
+            // If not coinbase restore prevout
+            if (!tx.IsCoinBase()) {
```
Is this special case actually needed? Seems like origcoins/oldcoins will be empty in this case anyway.

morcos replied:
It seems cleaner to me to not be messing with tx.vin of coinbases

ryanofsky commented on src/test/coins_tests.cpp:
```diff
@@ -227,77 +247,139 @@ BOOST_AUTO_TEST_CASE(updatecoins_simulation_test)
+            alltxs.insert(std::make_pair(tx.GetHash(),std::make_tuple(tx,undo,oldcoins)));
+
+            // Update the utxo set for future spends
+            utxoset.insert(tx.GetHash());
```
Suggestion: Maybe move the `utxoset` updates here and elsewhere next to the `result` updates, since `utxoset` is essentially just a list of the keys present in the `result` map. If somebody is changing one of these, they probably need to update the other too.

morcos replied:
hmm.. I'll move this above the alltxs insert, but I was trying to keep the order the same for each set of changes.

- update expected result
- modify cache the same way the production code does
(these first two things are what we will compare against each other)
- modify state used to run the test (which includes utxoset)

ryanofsky reviewed (approved):
Tested ACK 3572fea7dce3c20bfca237a53353cdf97c0f3669.

sipa commented:
utACK 743009682111aef5c4aab84c0cb5189e4afa4f06
Care to squash?

morcos commented:
Squashed
identical code to 7430096
only nits and minor changes to address test comments from 3572fea

TheBlueMatt commented on src/coins.cpp:
```diff
@@ -117,17 +117,30 @@ CCoinsModifier CCoinsViewCache::ModifyCoins(const uint256 &txid) {
-// This is to ensure that the historical duplicate coinbases before BIP30 was
-// in effect will still be properly overwritten when spent.
+/* ModifyNewCoins allows for faster coin modification when creating the new
+ * outputs from a transaction.  It assumes that BIP 30 applies and has already
```
I hate bip numbers...any chance you can also/or only state what this actually means?

TheBlueMatt replied:
(I think you did for 30, but not for 34)

TheBlueMatt replied:
You also didnt really mention that this is only true because we use a separate CCoinsViewCache for each block we (try to) connect.

TheBlueMatt commented:
While you're at it, can you add a comment to coins.cpp:200 (last else branch in BatchWrite) noting that it is very important that we are only swapping the coins itself, and keeping the flags set on the parent cache.

TheBlueMatt commented:
utACK code changes in c9270c96e2ff1676eb87a94ef9b872bea3653804, though I'd prefer more comments. Didnt review tests, but did spend a bunch of time looking at this with @morcos when he was deciding the best route to go for changes.

morcos commented:
@TheBlueMatt tell me if this is what you had in mind?  I'm not sure it's worth it, but don't mind squashing the additional comments if you want them.  The whole design of coins would be broken if we were copying all the flags over, but I tried to flag the confusing case.

TheBlueMatt commented:
Yea, looks good to me.

morcos commented:
squashed in additional comments

ryanofsky commented on src/coins.cpp:
```diff
@@ -117,17 +117,31 @@ CCoinsModifier CCoinsViewCache::ModifyCoins(const uint256 &txid) {
     assert(!hasModifier);
     std::pair<CCoinsMap::iterator, bool> ret = cacheCoins.insert(std::make_pair(txid, CCoinsCacheEntry()));
-    ret.first->second.coins.Clear();
     if (!coinbase) {
```
Just a suggestion, but I think another way you could write this section might be:

```
    if (!coinbase) {
        // New coins must not already exist.
        assert(ret.first->second.coins.IsPruned());

        if (!(ret.first->second.flags & CCoinsCacheEntry::DIRTY)) {
            // If the coin is known to be pruned (have no unspent outputs) in
            // the current view and the cache entry is not dirty, we know the
            // coin also must be pruned in the base view as well, so it is safe
            // to mark this fresh.
            ret.first->second.flags |= CCoinsCacheEntry::FRESH;
        }
    }
```

Differences from above:
- Assert is broader and more prominent.
- Comments are inline, which I think make this more understandable.
- FRESH flag will continue to be set in the case of the non-dirty existing cache entry.

morcos replied:
heh.  i'll want to think about that 100 times, but that sounds right to me

morcos commented:
dancing the dance of a clean rebase on the grave of main.cpp

will address @ryanofsky's suggestion separately

morcos commented:
@sipa @TheBlueMatt Can you guys take a look one more time at the new logic.  I'll squash to one commit after you ACK.

@fanquake Can you milestone for 0.14.0

sipa commented on src/coins.cpp:
```diff
@@ -121,24 +121,29 @@ CCoinsModifier CCoinsViewCache::ModifyCoins(const uint256 &txid) {
+            ret.first->second.flags |= CCoinsCacheEntry::FRESH;
         }
     }
     ret.first->second.coins.Clear();
```
This line could move into an else branch.

morcos replied:
Which line?  The `Clear()`?  Isn't that needed for the historical coinbase overwrite cases?

sipa reviewed (commented):
utACK e5f93783cd4c55a4a040719a9006e897e44509a3

sipa commented:
The branch above starts with an assert that coins.IsPruned(), so a Clear()
in that case would be a no-op (I think?).

morcos commented:
oh yes, confused myself..
i can change it if you think thats clearer, or a potential performance improvement?

ryanofsky commented:
Test fixes are here: https://github.com/ryanofsky/bitcoin/commits/pr/morcos-saferModifyNewCoins (in the two squash commits) if you see failures after rebasing.

sipa commented:
@morcos No strong opinion, just pointing it out. I don't think it will have a noticable performance impact.

morcos commented:
Rebased and squashed

Ignored @sipa's suggestion (sorry)
Included @ryanofsky's updates to the tests and switched the assert to throw std::logic_error so we could non-noisily test that code path.

Diff is here:
https://0bin.net/paste/RH2QCOc8C7lq505F#67QWes6s2d2bttLu6KXZqcKb7YJqEHzmY8pFv9+BV-M

ryanofsky reviewed (approved):
utACK b50cd7a67e71051db59199a4185e7c82b669c659

sipa commented:
utACK b50cd7a67e71051db59199a4185e7c82b669c659

sipa commented:
I ran the simulation test 1000x longer than the normal unit tests do.
